### ISSUE TITLE
Cascaded `NewType` causes unexpected pyright diagnostic

### ISSUE THREAD

**Issue by karolzlot** [microsoft/pyright#5197]
**Describe the bug**

This is valid in Pylance:
```python
from typing import NewType

Table = NewType("Table", str)
ChildTable = NewType("ChildTable", str)

def child_table(parent_table: Table) -> ChildTable:
    return ChildTable(f"{str(parent_table)}")
```

But this is not:

```python
from typing import NewType

Table = NewType("Table", str)
ChildTable = NewType("ChildTable", Table)

def child_table(parent_table: Table) -> ChildTable:
    return ChildTable(f"{str(parent_table)}")
```

![image](https://github.com/microsoft/pyright/assets/11590960/e180d21d-ac04-40f8-88ef-81df7e307e62)

![image](https://github.com/microsoft/pyright/assets/11590960/7a145607-5985-43a0-a127-3fafff54e054)

**To Reproduce**
Above.

**Expected behavior**
No diagnostic in both cases.

**Screenshots or Code**
Above.

**VS Code extension or command-line**
Pylance v2023.5.41

**Additional context**
n/a

**Comment by erictraut** [comment]
Yes, this is correct behavior. The `NewType` function was introduced in PEP 484 and is documented [here](https://peps.python.org/pep-0484/#newtype-helper-function). It creates a class with a constructor that accepts an argument of the specified base type. 

In your second example above, `ChildTable` is a class whose constructor accepts only a `Table` object as an argument, but you're attempting to pass a `str` value, which isn't compatible.

Not surprisingly, mypy generates the same error as pyright in this case.

**Comment by karolzlot** [comment]
@erictraut 

I want to point out inconsistency in how Pyright handles type checking for classes and NewType. 

According to PEP 484, NewType is meant to create type with almost zero runtime overhead and at the same time behave like subclass. Example of such subclass from PEP 484:

```python
class UserId(int):
    pass
```

But as demonstrated below, there are cases where NewType isn't behaving exactly like subclass from a type checking perspective.

This code is correct in pyright:

```python
class Table(str):
    pass

class ChildTable(Table):
    pass

def child_table1(parent_table: Table) -> ChildTable:
    return ChildTable(Table(str(parent_table)))  # correct in pyright

def child_table2(parent_table: Table) -> ChildTable:
    return ChildTable(str(parent_table))  # correct in pyright
```

And this is not:

```python
from typing import NewType

Table = NewType("Table", str)
ChildTable = NewType("ChildTable", Table)

def child_table1(parent_table: Table) -> ChildTable:
    return ChildTable(Table(str(parent_table)))  # correct in pyright

def child_table2(parent_table: Table) -> ChildTable:
    return ChildTable(str(parent_table))  # pyright error
```

Both examples are valid Python code.

**Comment by erictraut** [comment]
Please read my response above. Pyright is correct in generating an error here for the reasons I explained above. If you have further questions about how `NewType` is intended to work, feel free to post a question in the discussion thread.
